FormHelper Backed Enum support

ADmad opened the pull request:
- Coerce enum instances into their string/int values so that form widgets can property populate the input values.
- For table columns mapped to enum types auto set the form control type to `select` and auto generate the options list using the enum cases.

<!---

Please describe the reason for the changes you're proposing. If it is a large change, please open an issue to discuss first.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Handle backed enum to scalar value conversion.

This avoids error when for e.g. entity contains an enum object as a property value.

ADmad committed:
Auto generate select options for fields mapped to enum type.

ADmad commented:
Damn cross test contamination.

ADmad committed:
Update type

dereuromark commented on tests/test_app/TestApp/Model/Enum/ArticleStatusLabel.php:
```diff
@@ -0,0 +1,26 @@
+
+    public function label(): string
+    {
+        return 'Is ' . strtolower($this->name);
```
mb_ as best practice maybe?

ADmad replied:
This is a test class, not code that's going to be run for users so no need to be pedantic about it :)

ADmad commented:
@markstory There's some sequence of calls which causes the entry in `TypeFactory::$_builtTypes` for the enum type instance to be removed but related class name mapping in `TypeFactory::$_types` still remains, so when `TypeFactory::buildAll()` is called it tries to build the enum type instance itself which fails since it requires a 2nd argument.

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1336,6 +1353,24 @@ protected function _optionsOptions(string $fieldName, array $options): array
+
+        $values = [];
+        foreach ($enumClass::cases() as $case) {
+            $values[$case->value] = method_exists($case, 'label') ? $case->label() : $case->name;
```
Is the idea here that enums can implement `label()` to get custom formatting?

ADmad replied:
Yes, I have added a test case using the `ArticleStatusLabel` enum for the same.

markstory reviewed (commented):
I can look into the cross-test bleed through.

ADmad commented:
The following patch fixes the error and doesn't cause any other test failure
```diff
Database/TypeFactory.php b/src/Database/TypeFactory.php
index c5fe10bbff..5f05ebc77b 100644
--- a/src/Database/TypeFactory.php
+++ b/src/Database/TypeFactory.php
@@ -88,12 +88,11 @@ class TypeFactory
      */
     public static function buildAll(): array
     {
-        $result = [];
         foreach (static::$_types as $name => $type) {
-            $result[$name] = static::$_builtTypes[$name] ?? static::build($name);
+            static::$_builtTypes[$name] ??= static::build($name);
         }
 
-        return $result;
+        return static::$_builtTypes;
     }
 
     /**
@@ -106,7 +105,6 @@ class TypeFactory
     public static function set(string $name, TypeInterface $instance): void
     {
         static::$_builtTypes[$name] = $instance;
-        static::$_types[$name] = $instance::class;
     }
 
     /**
```

ADmad commented:
Setting `static::$_types[$name] = $instance::class` in `TypeFactory::set()` is principally incorrect since there's no guarantee that the factory can re-create the instance itself later, as is the case of enum types.

markstory commented:
> The following patch fixes the error and doesn't cause any other test failure

That looks like a solution to me as well. Beyond enum types, we can't assume that we'll be able to re-construct user defined types as well.

ADmad committed:
Fix db type instance creation error.

Don't add the class name to type map when setting a type instance using TypeFactory::set().
The factory can't know the constructor arguments of a custom types and can't rebuild their instance itself.

ADmad committed:
Suppress invalid phpstan error

othercorey commented:
Looks good! Nice working cleaning up the rough edges on enums.
